### ISSUE TITLE
ssreflect pollutes notation precedence space for colon

### ISSUE THREAD

**Issue by andrew-appel** [coq/coq#6078]
#### Version
8.7

#### Operating system
All

#### Description of the problem

I have an "only printing" notation that uses the colon : at level 26.  This works fine in Coq 8.6.1.  Apparently in Coq 8.7, I cannot use the colon at level 26, because ssreflect (which I do not want to use) reserves it at precedence level 100.   If it is true that the colon's precedence level is fixed at 100 because of something in a nonstandard library, I consider that a bug.  Especially that my notation is "only printing", so it can't have any effect on parsing!

See also 
https://github.com/PrincetonUniversity/VST/issues/111
https://github.com/PrincetonUniversity/VST/pull/112

**Comment by ejgallego** [comment]
If I understand the issue currently, you are indeed loading `ssreflect` in some of your dependencies, so even if you don't plan to use it, it is loaded.

If that is the case, I think it will hard to fix soon [unless ssreflect already has some special flag to help with this], the reason being how parsing rules are handled.

**Comment by ejgallego** [comment]
By the way, I think the problem should exist in previous versions of Coq + ssreflect, but need to double-check.

**Comment by gares** [comment]
From: http://math-comp.github.io/math-comp/htmldoc/mathcomp.ssreflect.ssreflect.html

     Allow overloading of the cast (x : T) syntax, put whitespace around the    
     ":" symbol to avoid lexical clashes (and for consistency with the parsing  
     precedence of the notation, which binds less tightly than application),    
     and put printing boxes that print the type of a long definition on a       
     separate line rather than force-fit it at the right margin.

```              
Notation "x : T" := (x : T)
  (at level 100, right associativity,
   format "'[hv' x '/ ' : T ']'") : core_scope.
```

I'm not very competent on notations, but i think application is at level 10, so both 100 and 26 should be fine.

Let me ping @ggonthier, @amahboubi and @CohenCyril to see if someone recalls why 100 exactly, and knows if lowering the level (e.g. to 26) would cause any issues in mathcomp. At the same time,
@andrew-appel, I could not find (in the two issues you linked) the reason why `:` is at level 26 and if raising its level to 100 would be an option.

The fact that that a notation/operator has a global level is clearly a problem (and maybe @ejgallego has an answer to that in his parsing-tables branch).
But while we wait for a better notation infrastructure, putting `:` at a level that works for both mathcomp and VST is probably doable. Also, moving the notation from `ssreflect.v` to `ssrfun.v` looks like another viable option to me (i.e. the offending notation would not be loaded by VST, that only requires `ssreflect.v` AFAICT).

**Comment by ppedrot** [comment]
> If I understand the issue currently, you are indeed loading ssreflect in some of your dependencies, so even if you don't plan to use it, it is loaded.

AFAIK, notations that are not imported have no effects, so it's even stronger than that: it means that ssreflect has been *imported* by one of the dependencies, not merely *required*. Which seems to me like a bug in the dependencies.

**Comment by ppedrot** [comment]
I had  a look at the notation implementation in `Metasyntax`, and I think it is easy to tweak the code so that only printing notations do not trigger such level conflicts. I'll write a quick PR to see it this is indeed workable.

**Comment by ggonthier** [comment]
The initial reason for overriding the cast notation was to enable the other overriding rule in ssreflect.v, for T : Type, which puts T in the %type scope, which gives a slightly more pleasant way of saying A * B : Type (vs. (A * B)%type). The printing rule for the generic cast was more like an added bonus. The intent for both rules was to be consistent with the native level for cast, which was 100 at the time this code was written.
   I’m surprised it’s ben changed by so much, because it is still documented as 100 in the 8.7 manual (end of section 12.1.3): changing that level requires adapting levels in all the comprehension notations, and since the new level is so low (binding tighter than arithmetic operators) require changing client code as well, e.g., [seq n + 1 | n <- iota 1 5] would require parentheses around the n + 1 expression (alongside spoiling the above mentioned A + B : Type notation).
    Georges

**Comment by herbelin** [comment]
Hi Georges,

> The initial reason for overriding the cast notation was to enable the other overriding rule in ssreflect.v, for `T : Type`, which puts T in the `%type` scope, which gives a slightly more pleasant way of saying `A * B : Type` (vs. `(A * B)%type`).

Is it possible that the other
```Notation "T : 'Type'" := (T%type : Type)  (at level 100, only parsing) : core_scope.```
work also without the overriding of the main cast notation?

> The printing rule for the generic cast was more like an added bonus. The intent for both rules was to be consistent with the native level for cast, which was 100 at the time this code was written.

The level of the native printing rule for cast is also 100, and associativity is the same, so I suspect that the primary reason for the overwriting was to add a space. The native printer adds spaces around the `:` since 8.5 (thanks to @gmalecha, see #3936), so could it be that the overwriting made in `ssreflect.v` is not necessary anymore to achieve this goal and that the rule can be removed without any negative impact on the habits of ssreflect users? (In any case, noone is proposing that 26 becomes the default level for cast.)

As a secondary remark, I like the idea of forcing the interpretation of `T` in `T:some_sort` to be done in `type_scope`. It is not excluded in principle that it introduce some incompatibility but I would be tempted to try to activate it by default (i.e w/o requiring a specific notation) since this is after all the intuitive expectation. At least, we could try to see how our test-suite react to such change, and if it works, that would be a simple but good example of how a good communication with ssreflect developers can serve the general interest of users.

Otherwise, about the VST issue, it is still not excluded, if I understood correctly, that there is an issue in the dependencies of `VST` in `ssreflect` or a Coq bug of `Require` wrt grammars.

As for the enhancement request that several notations with same "key" but different printing rules are supported, it also remains valid.

**Comment by herbelin** [comment]
PS: In the longer term (8.8), support for stand-alone `constr` parsers (PR #982, this [comment](https://github.com/coq/coq/pull/982#issuecomment-324617760)) shall give more flexibility in defining custom syntaxes for formalized languages such as Clight here (no interference with the main syntax on the choice of precedences or associativity).

**Comment by ggonthier** [comment]
Hi Hugo,
> Is it possible that the other
> Notation "T : 'Type'" := (T%type : Type) (at level 100, only parsing) : core_scope.
> work also without the overriding of the main cast notation?
> 
> The printing rule for the generic cast was more like an added bonus. The intent for both rules was to be consistent with the native level for cast, which was 100 at the time this code was written.
> 
> The level of the native printing rule for cast is also 100, and associativity is the same, so I suspect that the primary reason for the overwriting was to add a space. The native printer adds spaces around the : since 8.5 (thanks to @gmalecha <https://github.com/gmalecha>, see #3936 <https://github.com/coq/coq/issues/3936>), so could it be that the overwriting made in ssreflect.v is not necessary anymore to achieve this goal and that the rule can be removed without any negative impact on the habits of ssreflect users? (In any case, noone is proposing that 26 becomes the default level for cast.)
> 
It’s quite possible that the two overrides are linked conceptually rather than materially (i.e., once we’d picked up enough motivation to override the cast grammar, we just did both). If the overriding is now redundant, we should by all means remove it. Do note however that the generic cast override also adds a pretty-printing box, and this is much more important as it affects the readability of proof contexts for the reasons cited in the comment.
> As a secondary remark, I like the idea of forcing the interpretation of T in T:some_sort to be done in type_scope. It is not excluded in principle that it introduce some incompatibility but I would be tempted to try to activate it by default (i.e w/o requiring a specific notation) since this is after all the intuitive expectation. At least, we could try to see how our test-suite react to such change, and if it works, that would be a simple but good example of how a good communication with ssreflect developers can serve the general interest of users.
> 
By all means do experiment! Also, you could get a better implementation of this functionality while implementing this in system code:
  - imposing only transient (one-level) rather than deep scope override on T (unless there is now user syntax for that?)
  - generalising this to all type-bound scopes (so m + n : nat always interprets the + in nat_scope)
  - generalising this to implied casts, i.e., Definition AxB : Type := A * B does what you’d expect.
> Otherwise, about the VST issue, it is still not excluded, if I understood correctly, that there is an issue in the dependencies of VST in ssreflect or a Coq bug of Require wrt grammars.
> 
That seems likely; AFAIK unimproved grammar notations (in contrast to simple notations) have no effect. Perhaps some Import dependency of VST (improperly) exports ssreflect.

    Cheers,
Georges

**Comment by herbelin** [comment]
Hi Georges,

> It’s quite possible that the two overrides are linked conceptually rather than materially (i.e., once we’d picked up enough motivation to override the cast grammar, we just did both). If the overriding is now redundant, we should by all means remove it. Do note however that the generic cast override also adds a pretty-printing box, and this is much more important as it affects the readability of proof contexts for the reasons cited in the comment.

According to `git log`, there is a box for casts in the native printer since the very beginning of the new syntax (i.e. 8.0). So, "by all means remove it"?

> By all means do experiment!

Do you think that it is a good idea to propose @gares to do the experiment? (And if not, I can propose a PR for item 1 below.)

> Also, you could get a better implementation of this functionality while implementing this in system code:
>  - imposing only transient (one-level) rather than deep scope override on T (unless there is now user syntax for that?)

This is the default for those "ad hoc" internal forcing of a scope, so, that shall be easy.

>  - generalising this to all type-bound scopes (so m + n : nat always interprets the + in nat_scope)

Yes and no. Yes, this is convenient, but if we start in this direction, I don't know where to stop. Why `nat` and not a `_` easily inferable to be `nat`, why `nat` and not `id nat`, etc. I feel that internalize time is not the right place for this kind of feature (compare to pretyping). I'm not opposed but I wonder about what is the general opinion.

Is your motivation aesthetic (i.e compared to `(m+n)%nat`)? It is not that I find the notation `%` pretty illuminating, but a cast leaves a trace in the term, so if we want to favor the `t : T` in place of `t%TT` when type `T` is bound to a scope `TT`, this shall have impact beyond the purely surface language.

>  - generalising this to implied casts, i.e., Definition AxB : Type := A * B does what you’d expect.

It happens that this is already like this (since 8.5).

Best,

Hugo

**Comment by herbelin** [comment]
BTW, Georges, you are someone pretty sensitive to syntactic issues, so let me digress a bit. I tried a couple of time to motivate other people to integrate some of your syntax ideas that I find  cool, but with little success in getting feedback. For instance, I like `n.+1` but I did not perceive a strong will yet at providing it as soon as the initial state. On the other side, `if t is pat then u else v` seems to have received some support at the last Coq WG (while e.g. my efforts to convince OCaml to propose it failed). I don't think you are pretty motivated yourself with `existT`, but this is yet another syntax which looks rather ugly to me. Ideally, the most natural to me would be `(t,u)` but I don't know how to solve the conflict with `pair` (maybe `prod` and `sigT` should be non-generative after all, I think that this is also what @JasonGross suggested, and `prod` seen as a particular case of `sigT`). As a short-term plan, it could at least be `(t;u)` as used by @JasonGross. On another somewhat related question, what do you think about the idea of unifying `seq` and `list`? Best, Hugo

**Comment by andrew-appel** [comment]
Let me clarify, now that @ildyria has explained it better to me:

VST does not import or export ssreflect.  But some users of VST want to also use ssreflect, or vice versa.  It's only when they are combined that there is a problem.

**Comment by ggonthier** [comment]
> Le 6 nov. 2017 à 16:40, Hugo Herbelin <notifications@github.com> a écrit :
> 
> BTW, Georges, you are someone pretty sensitive to syntactic issues, so let me digress a bit. I tried a couple of time to motivate other people to integrate some of your syntax ideas that I find cool, but with little success in getting feedback. For instance, I like n.+1 but I did not perceive a strong will yet at providing it as soon as the initial state. On the other side, if t is pat then u else v seems to have received some support at the last Coq WG (while e.g. my efforts to convince OCaml to propose it failed). I don't think you are pretty motivated yourself with existT, but this is yet another syntax which looks rather ugly to me. Ideally, the most natural to me would be (t,u) but I don't know how to solve the conflict with pair (maybe prod and sigT should be non-generative after all, I think that this is also what @JasonGross <https://github.com/jasongross> suggested, and prod seen as a particular case of sigT). As a short-term plan, it could at least be (t;u) as used by @JasonGross <https://github.com/jasongross>. On another somewhat related question, what do you think about the idea of unifying seq and list? Best, Hugo
> 
> 

Thanks for trying! Regarding existT, math comp gets around some of the ugliness by using aliases for accessors and constructors (tag/tagged/Tagged), and we kind of get by with that because of the limited use we make of the generic sigT type - though I can clearly see how HoTT des might have different needs, and the (i; u) notation might be right (and may help deter users from overloading ‘;’, which wreaks havoc in the rest of the syntax). I still think there’s a big issue with the flow of information in type inference: I’d usually expect (i; u) to be untypeable unless the expected type for the expression is allowed to inform the expected type for u (specifically, provide the T in {i & T i} - indeed, in the Tagged alias i is implicit and T is explicit). I’ve already noted that this is a general problem with type inference of applied constants, and that it would help a lot if users could specify at which point the expected type of a application can be used to synthesise the type of remaining arguments, because the current default (« never ») is not always appropriate. Partly for this we tend not to use the generic sigT type, but define specialised sigT-like (co)inductive types - and we haven’t really found the need for a canonical structure reflecting that correspondance (as subType does for sig) simply because there aren’t enough theorems about sigT that could exploit this (I’ll grant you this may be very different in HoTT).

Concerning seq and list: they have long been unified (since 2007 if memory serves well): seq is merely an alias for list.Originally (for the 4 color theorem and in the very early math comp days) the types were different, because I’d used coercion to Funclass to denote list membership, and at the time this meant the argument to seq had to be an eqType structure to satisfy uniform inheritance. We’ve since moved to an explicit membership operator supporting type-based overloading, removing the need for a distinct type (uniform inheritance is sort of going away too). However I’ve found it difficult to replace seq by list, cat by append, size by length, etc, in the names of both auxiliary operations and related lemmas; in particular ‘l’ is a troublesome suffix to use, too easily confused with 1 and I. So, we’ve stuck with those aliases.
  There is a separate issue with the explicit list syntax: [] are the only ASCII brackets that aren’t reserved by the native Coq grammar, so it’s unreasonable to use them only for list syntax, especially as lists play a much smaller role in formalised theories as compared to functional programming. Our solution is to follow the leading ‘[‘ with a keyword or operator that specifies the kind of construction to follow, using ‘::' for explicit sequences, and ‘seq’ for sequence comprehensions. Perhaps there’s a better solution if one can tweak the parse engine (at least, one should be able to use the same separator for nil and cons terminated lists).

Best,
  Georges

**Comment by JasonGross** [comment]
>there aren’t enough theorems about sigT that could exploit this (I’ll grant you this may be very different in HoTT)

Indeed, we have about [650 lines of definitions and theorems](https://github.com/HoTT/HoTT/blob/master/theories/Types/Sigma.v) specific to `sigT`, basically all about equalities of sigma types.  It all generalized in principle to records (we in fact have [tactics for automatically proving equivalence between records and nested `sigT`s](https://github.com/HoTT/HoTT/blob/master/theories/Types/Record.v)), but it's currently impossible to quantify over all codes for records in the same way that one can quantify over "codes" for sigma types.  If Coq gains an anonymous record type constructor, then we could port these theorems to records in general.

**Comment by herbelin** [comment]
`sigT`: ok. Yes, there are a lot of improvements to do with unification but this is a difficult task. Often, when I (or others) try to make something a bit smarter, there is somewhere an example which takes benefit of the former weakness to bypass the weakness of another part of the algorithm. So, it looks like we shall have to make several changes globally, navigating through a sea of developments which may have innocently exploited oddities of the algorithm here and there. At least, what reassured me, is that Tobias and, even someone from Lean, which is however younger, also reported the same kind of difficulties at making their unification algorithm evolve in a compatible way.

`list` and `seq`: noted, too bad.

> Perhaps there’s a better solution if one can tweak the parse engine (at least, one should be able to use the same separator for nil and cons terminated lists).

Do you mean `;` instead of `,` in `[ :: x1 , x2 , .. , xn & s ]`? I would evaluate this as a little bug in camlp5 in the presence of (so-called) `LIST1`. If too rules have a `LIST1 entry ; other-prods` pattern, they are not shared and only the most recent one is considered.

I suspect this could be circumvented by factorizing the rules by hand and using the custom entry mechanism to come in 8.8, though, accordingly, this would decrease the readability of the declarations.

**Comment by mattam82** [comment]
About the directionality hints for type inference of applications I think this is perfectly doable while maintaining compatibility, and would subsume the special case of Program for constructors which exactly handles the sigT case, as far as I understand. I've been hoping to give it a go for some time but never got around to it (mainly because I just Set Program Mode and forget about it in my own developments). I guess its high time for a CEP.

**Comment by herbelin** [comment]
> According to git log, there is a box for casts in the native printer since the very beginning of the new syntax (i.e. 8.0). So, "by all means remove it"?
>
> >    By all means do experiment!
>
> Do you think that it is a good idea to propose @gares to do the experiment? (And if not, I can propose a PR for item 1 below.)

So, do we confirm an agreement on
- removing the `x : T` notation from `ssreflect.v`?
- experimenting the compatibility of hard-wiring in `constrintern` a `type_scope` on `T` for `T : some-sort`, and removing the `T : Type` and `P : Prop` notations in `ssreflect.v`?

If yes, I can do it, but I'd be happy also if one of you want to do it.

**Comment by JasonGross** [comment]
> experimenting the compatibility of hard-wiring in constrintern a type_scope on T for T : some-sort, and removing the T : Type and P : Prop notations in ssreflect.v?

Perhaps also experiment with something more general?  If there is a scope `s` bound to `T` (or to `Sortclass` when `T` is a sort), interpret `x : T` as `x%s : T`

**Comment by herbelin** [comment]
@JasonGross: why not but could you give some example where it is better to do `x:T` than `x%s`, knowing that the former adds a constructor in the type while the second is purely syntactical as shown in the following output:
```
Require Import ZArith.
Set Printing All.
Check fun x y z t => (x+y : nat, (z+t)%Z).
(* fun (x y : nat) (z t : Z) => @pair nat Z (Init.Nat.add x y : nat) (Z.add z t) *)
```
So is the motivation typing or aesthetic?

Or is it because you find that the distinction between a scope and a cast is after all not so clear and that something should be done to unify them?

Also, when you type `x + y : nat` do you expect the `nat` to stay in the term (as it is now), or do you expect it to be just a hint to be dropped after it is used?

**Comment by JasonGross** [comment]
> why not but could you give some example where it is better to do x:T than x%s

```coq
fun '((x, y) : myprod A B) => x
```

for `myprod` an inductive type like `prod` but with its own scope.  More generally, when the type is parameterized/indexed and the arguments to the type family need to be given anyway, and there's no more convenient place to give them, giving a scope by hand is redundant.

> Or is it because you find that the distinction between a scope and a cast is after all not so clear and that something should be done to unify them?

No, the distinction is very clear: scopes live only at parsing time to guide notations, and then disappear, types live in the terms and have no direct impact on parsing.  However, `Bind Scope` is extremely useful, and more generally, I frequently wish that Coq would automatically know what scope I want something to live in (in a principled way), and this is a move in that direction.  Actually, I have another additional wish in that directly wish, but I'll go open a separate issue for that.

> Also, when you type x + y : nat do you expect the nat to stay in the term 

I do, but maybe that's because I'm used to using Ltac match hacks to remove such casts when I want them gone.  I suppose the one place where I find this behavior annoying is when I want to change the type of a goal with evars to some other type and then refine with a thing of a convertible type, I frequently find it more convenient to write `refine (term : type)` and then remove the cast later if I have to, than to write ` match goal with |- ?G => unify G type end; refine term`.

**Comment by herbelin** [comment]
OK, I see. Unless someone opposes, I may try to consider experimenting this wish voiced twice (+ one :+1:).

**Comment by herbelin** [comment]
For the record, the redefinition of the cast notation in `ssreflect.v` has an impact also on the interpretation of notations: the body of the cast is interpreted in the empty local scope with the overloading, while it is interpreted in the local scope active at the time of the cast in the native rule.

The proposal to have the body interpreted in a scope triggered by the type (and none if the type is not bound to a scope), implemented in #6134, goes in the direction of a merge of behaviors in 8.8. For 8.7.1, would it be possible that the notations redefined in `ssreflect.v` are moved to a file of math-comp rather than being in the `ssreflect.v` of the plugin?

**Comment by herbelin** [comment]
For the record:

> Do you mean ; instead of , in [ :: x1 , x2 , .. , xn & s ]? I would evaluate this as a little bug in camlp5 in the presence of (so-called) LIST1. If two rules have a LIST1 entry ; other-prods pattern, they are not shared and only the most recent one is considered.

With the help of @roglo, I added a workaround to force the factorization of such rules when the separator is a single token. This is PR #6167.

The reason for non-factorization was that the parsing of separators was implemented using a grammar sub-action, that the latter is a function, and that factorization does not look at the body of functions. In the case of an atomic separator we can avoid the use a sub-action by directly returning the token. See the discussion at PR #6167 for further possible workarounds in the general case, if needed.

**Comment by Zimmi48** [comment]
> For 8.7.1, would it be possible that the notations redefined in ssreflect.v are moved to a file of math-comp rather than being in the ssreflect.v of the plugin?

I really think that fixing this in 8.7.1 should not be viewed as a priority (unless it is a regression, but from what I understand, it isn't). Let's remind that the release of 8.8.0 is expected in just five months.

So I would argue against the proposed move. First of all, we've seen that the removal of the notation causes an incompatibility in math-comp, but that could also be the case in other libraries that depend on SSReflect. Second the code of SSReflect is now tightly coupled to the Coq version, but that's not the case for math-comp, so moving the notation there would mean keeping it for much longer than necessary.

**Comment by herbelin** [comment]
@Zimmi48: As said in the parallel discussion at #6133, I agree with targetting 8.8.0 as it is actually not a regression.

**Comment by ildyria** [comment]
Just a reminder that this issue is sadly still alive. :(

**Comment by gares** [comment]
I tried to reconstruct the status, but I think @herbelin knows better.

I think #6134 is stuck, and it is necessary in order to remove the cast notation from ssreflect.v (I quite don't get if it is sufficient).  @herbelin is this correct? PR #6167 (merged now) is an accessory, not related to this issue.

FTR moving the problematic notation to math-comp is not solving this issue, since it appears in users of math-comp + VST.

**Comment by andrew-appel** [comment]
As far as I can tell, this issue is now solved in Coq 8.11.0.  Perhaps that's because ssreflect no longer uses a colon notation?  Shall we close this issue?

**Comment by gares** [comment]
I guess so.
